### PR TITLE
Mypy plugin

### LINKED ISSUE #460: mypy plugin
# Feature Request

I believe that this project could benefit from having a small mypy plugin built for it. Errors like #400 or #280 could be addressed in this plugin.

## References

- [Mypy blog post about plugins](http://mypy-lang.blogspot.com/2019/03/extending-mypy-with-plugins.html)
- [django-mypy-plugin](https://github.com/mkurnikov/django-stubs/tree/master/mypy_django_plugin)
- [SQLAlchemy-stubs](https://github.com/dropbox/sqlalchemy-stubs/blob/master/sqlmypy.py)
- [dataclasses mypy plugin](https://github.com/python/mypy/blob/master/mypy/plugins/dataclasses.py)

### PR DESCRIPTION
**Pull request by dmontagu**
## Change Summary

Add a mypy plugin with a number of features; see the added docs for more details.

Other notable improvements for pydantic *developers*:
* The mypy tests have been reorganized to be easier to work on
    * Eventually, it will probably make sense to better organize individual tests, but
    this seemed like the most practical solution for now
* Thanks to the reorganization, the mypy tests now run after all others (I think this is preferable)
* The mypy tests and codebase check now run much faster on repeated execution because they no longer invalidate each other's caches
    * The tests now make use of separate cache directories by plugin settings, all contained as subfolders of the previously used `.mypy_cache` folder.
    * Because they are subfolders of `.mypy_cache`, the same `make clean` works to clean up all of the cache directories

## Related issue number

Closes #460 

## Checklist

* [x] Unit tests for the changes exist
* [x] Tests pass on CI and coverage remains at 100%
* [x] Documentation reflects the changes where applicable
* [x] change description file added to `changes/`, 
  see [changes/README.md](https://github.com/samuelcolvin/pydantic/blob/master/changes/README.md) for details
  on the format

### CODE REVIEW

**Comment by samuelcolvin** [comment]
In principle this looks great, but as you say let's defer it until 1.1 - it shouldn't cause any backwards incompatibility issues I assume?

File: setup.cfg [hunk 1]
```diff
@@ -41,6 +42,8 @@ warn_redundant_casts = True
 warn_unused_ignores = True
 disallow_any_generics = True
 check_untyped_defs = True
+# Change to `plugins = pydantic.mypy:strict` to type-check and remove `**kwargs` from BaseModel.__init__:
+plugins = pydantic.mypy
```
Comment (dmontagu) [inline setup.cfg]: > Very dumb/simple question: how is this plugin distributed? As part of pydantic or as a separate package?

I wasn't sure  of the answer, so I tried this (based on the mypy docs) and it just worked. If you change `pydantic.mypy` to `pydantic.mypy:strict`, it will fail the external mypy test (since the external test involves inputs getting parsed, so aren't the right types in the input to the `__init__`); this is my proof it is being used :)

Comment (samuelcolvin) [reply setup.cfg]: :+1:

**Comment by dmontagu** [comment]
You can always just turn the plugin off (by not incluiding it in `mypy.ini`/`setup.cfg`), so it is strictly opt-in. So I don't think there are backwards compatibility issues.

**Comment by dmontagu** [comment]
For what it's worth, if anyone wants this in the interim, you can just copy/paste the contents of the pydantic/mypy.py file into your own project and point mypy at it, and it should just work. (Doesn't require any changes to pydantic.)

It would be nice to be able to distribute it with the library though.

**Comment by dmontagu** [comment]
After some painful edge case exploration and mypy debugging, I've gotten the test coverage of the plugin itself up to 100% (modulo two minor no-covers related to import edge cases, where I've just copied the logic from the official dataclasses plugin).

Still needs:
- [ ] docs
- [ ] support for type checking the `construct` method

I think it might make sense to (eventually) support other config settings and/or finer-grained strictness controls, but I'm more inclined to just release this and see what features people actually request.

File: tests/test_mypy.py [hunk 2]
```diff
@@ -1,61 +0,0 @@
-import importlib
```
Comment (dmontagu) [inline tests/test_mypy.py]: Moving this file both 1) makes it easier to dev the mypy tests since everything is closer together, and 2) causes the mypy tests to execute last. Given they are a little slow to execute, this seems preferable to me.

**Comment by dmontagu** [comment]
@samuelcolvin I think this is basically done! I added docs and updated the description at the top with some more detail.

The docs cover basically everything the plugin adds.

The tests are a little haphazard organized, basically just dumped into the big `plugin-success.py` and `plugin-fail.py` files; we can fix that eventually but I personally think this is good enough for a first release. I'm happy to put more effort into cleaning it up though if you want that to happen before a merge.

Special note: these changes should make running the mypy tests and mypy check faster on subsequent runs due to better caching (even though there are now more tests).

**Comment by dmontagu** [comment]
Also, @koxudaxi if you have a chance to review I think you might find it interesting and/or have some inputs.

**Comment by samuelcolvin** [comment]
I'm away from my computer but will look on Tuesday if not before.

Are we caching the mypy cache directory (sorry can't see on my phone)?

**Comment by dmontagu** [comment]
@samuelcolvin Sounds good, no rush.

I’m not sure what you mean by “are we caching the mypy cache directory” — could you clarify?

I put the test caches inside the default one (used during checks). The structure of the caching directories is such that this doesn’t cause any conflicts.

**Comment by samuelcolvin** [comment]
Sorry, I mean there's a `.mypy_cache` or simpler directory. Have we added that to the list of cached directories in Travis?

**Comment by dmontagu** [comment]
Yeah, the directory is `.mypy_cache`. I didn’t know you could cache directories in Travis. Actually I don’t know much about Travis in general. I’m not at my computer now either but can try to figure it out when I can if you’d like.

**Comment by samuelcolvin** [comment]
Well, might make things faster. Don't worry if it's a pain.

File: docs/mypy_plugin.md [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+  keyword arguments.
+* If `Config.allow_population_by_field_name=True`, the generated signature will use the field names,
+  rather than aliases.
+* For subclasses of `BaseSettings`, all fields are treated as optional since they may be read from the environment.
+* If `Config.extra="forbid"` and you don't make use of dynamically-determined aliases, the generated signature
+  will not allow unexpected inputs.
+* **Optional:** If the `init_allow_extra` **plugin setting** (see below) is set to `False`, unexpected inputs to
+  `__init__` will raise errors even if `Config.extra` is not `"forbid"`.
+* **Optional:** If the `init_typed` **plugin setting** (see below) is set to `True`, the generated signature
+  will use the types of the model fields (otherwise they will be annotated as `Any` to allow parsing).
```
[the remark this answers is not in the document]
Comment (dmontagu) [reply docs/mypy_plugin.md]: Sounds good! Let me know if there are features the pycharm plugin handles that are conspicuously missing here and I can try to add them. (E.g., validation of the signature of `validator`-decorated methods.) I think it makes sense to release before getting *everything* covered (as long as there aren't obvious bugs in provided functionality), but it would be good to track.

File: docs/mypy_plugin.md [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+ 
+#### Generate a typed signature for `Model.construct`
+* This method is a (potentially much faster) alternative to `__init__` when input data is known to be valid,
+  and does not need to be  parsed
```
[the remark this answers is not in the document]
Comment (dmontagu) [reply docs/mypy_plugin.md]: Thanks for catching this!

File: docs/mypy_plugin.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (koxudaxi) [reply docs/mypy_plugin.md]: I have read all your mypy code.
Does show this plugin an error when access fields on a class method? 
eg: 
```python
class Model(BaseModel):
  name: str = 'abc'
  @classmethod
  def get_default_name(cls):
       return cls.name   # <- here
```
Otherwise, It looks good :)

[the remark this answers is not in the document]
Comment (dmontagu) [reply docs/mypy_plugin.md]: That’s an interesting catch!

I doubt it shows an error, but at the same time, I’m not sure mypy would typically show an error there, eg if using a dataclass. But I agree it could be a useful error to catch, especially for newcomers to pydantic (and especially in validators).

I’m not sure how I’d override that behavior but I’ll think about it. Any insight would be appreciated though!

Either way I don’t think this needs to block a first release.

File: tests/mypy/test_mypy.py [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+
+# This ensures mypy can find the test files, no matter where tests are run from:
+os.chdir(Path(__file__).parent.parent.parent)
+
+# You can change the following variable to True during development to overwrite expected output with generated output
+GENERATE = False
+
+cases = [
+    ('mypy-plugin.ini', 'plugin_success.py', None),
+    ('mypy-plugin.ini', 'plugin_fail.py', 'plugin-fail.txt'),
```
Comment (samuelcolvin) [inline tests/mypy/test_mypy.py]: are you sure these are really easier as separate files? Surely easier to have the expected output right here with the test even if some are very verbose?

File: docs/mypy_plugin.md [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+if your field names or types change.
+
+There are other benefits too! See below for more details.
+
+### Plugin Capabilities
+
+#### Generate a signature for `Model.__init__`
+* Any required fields that don't have dynamically-determined aliases will be included as required
+  keyword arguments.
+* If `Config.allow_population_by_field_name=True`, the generated signature will use the field names,
```
[the remark this answers is not in the document]
Comment (dmontagu) [reply docs/mypy_plugin.md]: The way mypy works in this context is that you can basically programatically generate a method definition, and then it is type-checked as though it had been written in code. So, you can't really encode the alias-handling logic in a way mypy supports -- it has to be checked like a normal python method.

I can't imagine a circumstance under which you would have `allow_population_by_field_name = True`, and also prefer to use the *alias* in your code. It seems to me the vast majority of usage of this feature is to have a pythonic name for use in code, and a non-pythonic (e.g., camel case) alias.

I'm not saying it's not possible that someone would prefer to use the alias in their code, but then I just don't understand why they wouldn't have named the variable that.

[the remark this answers is not in the document]
Comment (dmontagu) [reply docs/mypy_plugin.md]: We could set it up so that whenever mypy can determine the alias, that is what is used, but I think that is also problematic: for example, how should we handle cases where the alias contains a hyphen? In the end, it seems to me that the 1) highest-consistency, 2) highest user-convenience, and 3) easiest to implement solution is to just use the non-alias names whenever that is an option.

[the remark this answers is not in the document]
Comment (dmontagu) [reply docs/mypy_plugin.md]: Theoretically, I guess we could create overloaded signatures with both versions of the argument where possible, but that would lead to 2^n overloads where n is the number of aliases, so the error messages would get completely out of control.

File: docs/mypy_plugin.md [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+  checks or raise spurious errors.
+
+### Enabling the Plugin
+
+To enable the plugin, just add `pydantic.mypy` to the list of plugins in your mypy config file.
+
+For example, a `mypy.ini` file with the pydantic mypy plugin enabled might look like this:
+```ini
+[mypy]
+plugins = pydantic.mypy
```
[the remark this answers is not in the document]
Comment (dmontagu) [reply docs/mypy_plugin.md]: I wanted to introduce the complexity one bit at a time, and you don't need to worry about the `[pydantic-mypy]` section unless you want to configure special options.

I see two options: 1) I could add a link to the section below with the `[pydantic-mypy]` section, or 2) I could add the section here and set the values to the defaults.

File: docs/mypy_plugin.md [hunk 8]
```diff
@@ -0,0 +1,137 @@
...
+
+# for strict mypy: (this is the tricky one :-))
+disallow_untyped_defs = True
+```
+
+### Plugin Settings
+
+The plugin offers a few optional strictness flags if you want even stronger checks:
+
+* `allow_required_dynamic_aliases` (default: `True`)
```
[the remark this answers is not in the document]
Comment (dmontagu) [reply docs/mypy_plugin.md]: I think that makes sense

File: pydantic/mypy.py [hunk 9]
```diff
@@ -0,0 +1,624 @@
...
+    TypeType,
+    TypeVarDef,
+    TypeVarType,
+    UnionType,
+)
+from mypy.typevars import fill_typevars
+from mypy.util import get_unique_redefinition_name
+
+T = TypeVar('T')
+CB = Optional[Callable[[T], None]]
```
[the remark this answers is not in the document]
Comment (dmontagu) [reply pydantic/mypy.py]: Probably not 😅 most of the strangeness like this that you are seeing here and below are due to the fact that I started from the [attrs](https://github.com/python/mypy/blob/master/mypy/plugins/attrs.py) / [dataclasses](https://github.com/python/mypy/blob/master/mypy/plugins/dataclasses.py) mypy plugins.

File: pydantic/mypy.py [hunk 10]
```diff
@@ -0,0 +1,624 @@
...
+
+CONFIGFILE_KEY = 'pydantic-mypy'
+METADATA_KEY = 'pydantic-mypy-metadata'
+BASEMODEL_FULLNAME = 'pydantic.main.BaseModel'
+BASESETTINGS_FULLNAME = 'pydantic.env_settings.BaseSettings'
+FIELD_FULLNAME = 'pydantic.fields.Field'
+
+
+def plugin(version: str) -> 'TypingType[Plugin]':
+    """`version` is the mypy version string"""
```
[the remark this answers is not in the document]
Comment (dmontagu) [reply pydantic/mypy.py]: Taken verbatim from one of the other plugins; I'll make the docstring format more consistent

File: pydantic/mypy.py [hunk 11]
```diff
@@ -0,0 +1,624 @@
...
+    def __init__(self, options: Options) -> None:
+        self.plugin_config = PydanticPluginConfig(options)
+        super().__init__(options)
+
+    def get_base_class_hook(self, fullname: str) -> 'CB[ClassDefContext]':
+        sym = self.lookup_fully_qualified(fullname)
+        if sym and isinstance(sym.node, TypeInfo):  # pragma: no branch
+            # No branching may occur if the mypy cache has not been cleared
+            if is_pydantic_model(sym.node):
+                return partial(pydantic_model_class_maker_callback, plugin_config=self.plugin_config)
```
[the remark this answers is not in the document]
Comment (dmontagu) [reply pydantic/mypy.py]: sounds good, another awkwardness due to originally following the structure of the other plugins

File: pydantic/mypy.py [hunk 12]
```diff
@@ -0,0 +1,624 @@
...
+        return None
+
+    def get_method_hook(self, fullname: str) -> Optional[Callable[[MethodContext], Type]]:
+        if fullname.endswith('.from_orm'):
+            return from_orm_callback
+        return None
+
+
+class PydanticPluginConfig:
+    allow_required_dynamic_aliases: bool
```
[the remark this answers is not in the document]
Comment (dmontagu) [reply pydantic/mypy.py]: Yes, good catch

File: pydantic/mypy.py [hunk 13]
```diff
@@ -0,0 +1,624 @@
...
+    allow_untyped_fields: bool
+    init_allow_extra: bool
+    init_typed: bool
+
+    def __init__(self, options: Options) -> None:
+        default_plugin_settings: Dict[str, bool] = dict(
```
[the remark this answers is not in the document]
Comment (dmontagu) [reply pydantic/mypy.py]: yeah, I was struggling to come up with a clean way to handle this. I'll try that

File: pydantic/mypy.py [hunk 14]
```diff
@@ -0,0 +1,624 @@
...
+
+def from_orm_callback(ctx: MethodContext) -> Type:
+    """Raise an error if orm_mode is not enabled"""
+    model_type: Instance
+    if isinstance(ctx.type, CallableType) and isinstance(ctx.type.ret_type, Instance):
+        model_type = ctx.type.ret_type  # called on the class
+    elif isinstance(ctx.type, Instance):
+        model_type = ctx.type  # called on an instance (unusual, but still valid)
+    else:  # pragma: no cover
+        # I'm not sure if there is a way to get to this branch, but if so, it is very rare
```
[the remark this answers is not in the document]
Comment (dmontagu) [reply pydantic/mypy.py]: One of the mypy devs told me it can be something else, but didn't tell me what else it can be! I'll add a little convenience function that says please create a github issue to let us know you ran into this fork.

File: pydantic/mypy.py [hunk 15]
```diff
@@ -0,0 +1,624 @@
...
+    return ctx.default_return_type
+
+
+class PydanticModelTransformer:
+    _fields_set_argument: Optional[Argument] = None
+
+    def __init__(self, ctx: ClassDefContext, plugin_config: PydanticPluginConfig) -> None:
+        self._ctx = ctx
+        self.plugin_config = plugin_config
+        self.model_config_fields = {
```
[the remark this answers is not in the document]
Comment (dmontagu) [reply pydantic/mypy.py]: yep

File: pydantic/mypy.py [hunk 16]
```diff
@@ -0,0 +1,624 @@
...
+            'extra',
+            'allow_mutation',
+            'orm_mode',
+            'allow_population_by_field_name',
+            'alias_generator',
+        }
+
+    def transform(self) -> None:
```
[the remark this answers is not in the document]
Comment (dmontagu) [reply pydantic/mypy.py]: sounds good

File: pydantic/mypy.py [hunk 17]
```diff
@@ -0,0 +1,624 @@
...
+        ctx = self._ctx
+        info = self._ctx.cls.info
+
+        config = self.collect_config()
+        attributes = self.collect_fields(config)
```
[the remark this answers is not in the document]
Comment (dmontagu) [reply pydantic/mypy.py]: yeah, it's a place I forgot to change the name to field; I'll set it to that

File: pydantic/mypy.py [hunk 18]
```diff
@@ -0,0 +1,624 @@
...
+        is_settings = self.is_settings()
+        self.add_initializer(attributes, config, is_settings)
+        self.add_construct_method(attributes, is_settings)
+        if config.allow_mutation is False:
+            self.set_frozen(attributes, frozen=True)
+        else:
+            self.set_frozen(attributes, frozen=False)
+
+        info.metadata[METADATA_KEY] = {
+            'attributes': OrderedDict((attr.name, attr.serialize()) for attr in attributes),
```
[the remark this answers is not in the document]
Comment (dmontagu) [reply pydantic/mypy.py]: I don't think so, got it from another plugin

File: pydantic/mypy.py [hunk 19]
```diff
@@ -0,0 +1,624 @@
...
+    def is_settings(self) -> bool:
+        for info in self._ctx.cls.info.mro[:-1]:  # 0 is the current class, -1 is object
+            if info.fullname() == BASESETTINGS_FULLNAME:
+                return True
+        return False
+
+    def add_initializer(self, fields: List['PydanticModelField'], config: 'ModelConfigData', is_settings: bool) -> None:
+        ctx = self._ctx
+        typed = self.plugin_config.init_typed
+        use_alias = config.allow_population_by_field_name is not True
```
[the remark this answers is not in the document]
Comment (dmontagu) [reply pydantic/mypy.py]: In general it is not possible to statically determine the truthiness of the config settings, but even to handle a cases like `= 1` would add a painful amount of additional code. So I actually set things up so that you get a separate error if you set the config to anything that isn't a boolean.

I originally wrote the config to have `Optional[bool]` values so I could tell whether they were set, but I ultimately made it always bool. So I think I can safely change the `is not True` to `is False`.

File: pydantic/mypy.py [hunk 20]
```diff
@@ -0,0 +1,624 @@
...
+        force_all_optional = is_settings or bool(
+            config.has_alias_generator and not config.allow_population_by_field_name
```
[the remark this answers is not in the document]
Comment (dmontagu) [reply pydantic/mypy.py]: If you allow population by field name, we can just default to the field names. If you don't allow population by field name, and don't have an alias generator, we can check each field individually for aliases set in the `Field` call. If you don't allow population by field name and *do* have an alias generator, we won't know the name of *any* of the required arguments, so they have to all become optional.

Handling this (relatively common) case properly and consistently is one of the main reasons I generally prioritized the use of the non-alias version of the argument when possible.

File: pydantic/mypy.py [hunk 21]
```diff
@@ -0,0 +1,624 @@
...
+                and isinstance(expr.callee, RefExpr)
+                and expr.callee.fullname == FIELD_FULLNAME
+            ):
+                arg_is_ellipsis = len(expr.args) > 0 and type(expr.args[0]) is EllipsisExpr
+            else:
+                arg_is_ellipsis = isinstance(expr, EllipsisExpr)
+            return arg_is_ellipsis
+        # I believe TempNode means annotation-only
+        value_type = cls.info[lhs.name].type
+        if type(value_type) is UnionType:
```
[the remark this answers is not in the document]
Comment (dmontagu) [reply pydantic/mypy.py]: I probably adapted this from a copy-pasted line following this convention. I'll switch it

File: pydantic/mypy.py [hunk 22]
```diff
@@ -0,0 +1,624 @@
...
+            if field.has_dynamic_alias:
+                return True
+        if has_alias_generator:
+            for field in fields:
+                if field.alias is None:
+                    return True
+        return False
+
+
+class PydanticModelField:
```
[the remark this answers is not in the document]
Comment (dmontagu) [reply pydantic/mypy.py]: I wanted to do that, but it would break compatibility with 3.6 without the dataclasses backport. Let me know if you would want that despite the restriction it would add

File: pydantic/mypy.py [hunk 23]
```diff
@@ -0,0 +1,624 @@
...
+            'line': self.line,
+            'column': self.column,
+        }
+
+    @classmethod
+    def deserialize(cls, info: TypeInfo, data: JsonDict) -> 'PydanticModelField':
+        return cls(**data)
+
+
+class ModelConfigData:
```
[the remark this answers is not in the document]
Comment (dmontagu) [reply pydantic/mypy.py]: Same issue noted above.

File: pydantic/mypy.py [hunk 24]
```diff
@@ -0,0 +1,624 @@
...
+
+
+ERROR_ORM = ErrorCode('pydantic-orm', 'Invalid from_orm call', 'Pydantic')
+ERROR_CONFIG = ErrorCode('pydantic-config', 'Invalid config value', 'Pydantic')
+ERROR_ALIAS = ErrorCode('pydantic-alias', 'Dynamic alias disallowed', 'Pydantic')
+ERROR_UNTYPED = ErrorCode('pydantic-field', 'Untyped field disallowed', 'Pydantic')
+
+
+def error_from_orm(model_name: str, api: CheckerPluginInterface, context: Context) -> None:
+    api.fail(f'"{model_name}" does not have orm_mode=True', context, code=ERROR_ORM)
```
[the remark this answers is not in the document]
Comment (dmontagu) [reply pydantic/mypy.py]: I wanted there to be an obvious place where all the errors lived to make it easier to consistently change their format, update the error code, etc. I can change this if you'd prefer, but I do think there's a benefit to organizing things like this.

[the remark this answers is not in the document]
Comment (dmontagu) [reply pydantic/mypy.py]: I'd also be happier to inline this (and the one below) once we are a little more confident we won't be iterating on the error messages / api. (That may well be before this is merged 😅 in which case I can inline it then if preferred.)

File: tests/mypy/test_mypy.py (hunk 5 above)
Comment (dmontagu) [reply tests/mypy/test_mypy.py]: It's *extremely* hard to iterate on it without a mechanism for generating the expected output. By putting it in a file, I can set `GENERATE=True` above and it will write the file with the actual output, and I can look at the git diffs to see if things are as desired.

I initially tried having the results in-line here, but the problem is that a one line changes *all* the line numbers. And the line numbers actually do matter, so I think we shouldn't just strip them from the expected output. But then every time I tweak a line I had to manually compare all the output lines and fix the line numbers. Then I'd make another minor change and have to do it all again.

I think there might be better ways to manage mypy plugin tests (might be good to look at how the sqlalchemy or django plugins do this). But I thought they might ultimately rely on calling mypy on a separate file for each test, and I was specifically trying to minimize the amount of time spent executing mypy tests during normal test runs, and I think there is a substantial amount of overhead for each file analyzed. I tried to get all the tests into as few files as possible to keep the tests running fast.

File: pydantic/mypy.py [hunk 25]
```diff
@@ -0,0 +1,624 @@
...
+            ):
+                arg_is_ellipsis = len(expr.args) > 0 and type(expr.args[0]) is EllipsisExpr
+            else:
+                arg_is_ellipsis = isinstance(expr, EllipsisExpr)
+            return arg_is_ellipsis
+        # I believe TempNode means annotation-only
+        value_type = cls.info[lhs.name].type
+        if type(value_type) is UnionType:
+            item_types = [type(item) for item in value_type.items]  # type: ignore
+            if NoneType in item_types:
```
[the remark this answers is not in the document]
Comment (dmontagu) [reply pydantic/mypy.py]: This `NoneType` is `mypy.types.NoneType`, not `type(None)` (it *is* different, not just an alias for `type(None)`, and is related to how they track variable types).

File: pydantic/mypy.py [hunk 26]
```diff
@@ -0,0 +1,624 @@
...
+def error_invalid_config_value(name: str, api: SemanticAnalyzerPluginInterface, context: Context) -> None:
+    api.fail(f'Invalid value for "Config.{name}"', context, code=ERROR_CONFIG)
+
+
+def error_required_dynamic_aliases(api: SemanticAnalyzerPluginInterface, context: Context) -> None:
+    api.fail('Required dynamic aliases disallowed', context, code=ERROR_ALIAS)
+
+
+def error_untyped_fields(api: SemanticAnalyzerPluginInterface, context: Context) -> None:
+    api.fail('Untyped fields disallowed', context, code=ERROR_UNTYPED)
```
[the remark this answers is not in the document]
Comment (dmontagu) [reply pydantic/mypy.py]: Same note as above

**Comment by dmontagu** [comment]
@samuelcolvin I think I've incorporated all your feedback on the non-docs files, or at least responded above where I had concerns.

I renamed the plugin config settings so they could all be `False` by default (but basically have the same meaning).

I'm going to wait until we reach consensus on the non-docs stuff before updating the docs.

**Comment by dmontagu** [comment]
It looks like [this](https://github.com/typeddjango/pytest-mypy-plugins) is what django-stubs uses.

Here's a blog article discussing: https://sobolevn.me/2019/08/testing-mypy-types

It looks like a massive improvement over what is used in this PR, but 1) I'm not sure how fast it would run, and 2) I'm not sure how maintained it is.

File: docs/mypy_plugin.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (samuelcolvin) [reply docs/mypy_plugin.md]: I was just thinking of creating a signature something like:

```py
def __init__(self, foo_bar: str = None, fooBar: str = None, user_id: int = None, userId: int = None):
    ...
```
But as i wrote that I realised it wouldn't work for required fields.

Let's stick to the current behaviour.

File: docs/mypy_plugin.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (samuelcolvin) [reply docs/mypy_plugin.md]: make sense.

Maybe:
* remove the `follow_imports...disallow_untyped_defs = True` part since that confuses this simple example
* add a link to `/usage/mypy/` saying something like "go here for more details on mypy usage"

File: pydantic/mypy.py (hunk 24 above)
[the remark this answers is not in the document]
Comment (samuelcolvin) [reply pydantic/mypy.py]: fine by me to keep it like this.

File: pydantic/mypy.py (hunk 20 above)
[the remark this answers is not in the document]
Comment (samuelcolvin) [reply pydantic/mypy.py]: okay

File: pydantic/mypy.py (hunk 25 above)
[the remark this answers is not in the document]
Comment (samuelcolvin) [reply pydantic/mypy.py]: okay.

File: pydantic/mypy.py (hunk 19 above)
[the remark this answers is not in the document]
Comment (samuelcolvin) [reply pydantic/mypy.py]: Okay, this is good.

File: pydantic/mypy.py (hunk 22 above)
[the remark this answers is not in the document]
Comment (samuelcolvin) [reply pydantic/mypy.py]: But the backport of dataclasses is a required dependency.

Unless there's some reasonable way to use this plugin without installing pydantic (and therefore dataclasses), I think it's fine to use dataclasses.

Up to you though.

File: tests/mypy/test_mypy.py [hunk 27]
```diff
@@ -14,7 +14,7 @@
 os.chdir(Path(__file__).parent.parent.parent)
 
 # You can change the following variable to True during development to overwrite expected output with generated output
-GENERATE = False
+GENERATE = True
```
Comment (samuelcolvin) [inline tests/mypy/test_mypy.py]: should this be `True`? If so maybe, you need to edit the comment above?

**Comment by samuelcolvin** [comment]
Looking good. 

Apart from the docs and a few pretty trivial potential fixes, I think this is more or less ready.

File: pydantic/mypy.py (hunk 22 above)
[the remark this answers is not in the document]
Comment (dmontagu) [reply pydantic/mypy.py]: I see, I didn't realize that. Then yeah I think dataclasses is better for sure.

**Comment by dmontagu** [comment]
I tried using dataclasses but it caused the cythonized builds to fail due to the `__init__` method not getting created properly.

Maybe there's a workaround, but since it's not really giving much benefit here I figured hand-writing the __init__ function was fine.

**Comment by dmontagu** [comment]
Also, I *believe* I have now addressed all your concerns with regard to the docs.

**Comment by samuelcolvin** [comment]
this is awesome. 

Thank you very much!
